druntime/src/core/internal/array/capacity.d

```diff
@@ -8,11 +8,83 @@
   Source: $(DRUNTIMESRC core/internal/_array/_capacity.d)
 */
 module core.internal.array.capacity;
+import core.exception : onFinalizeError;
 
 // HACK: `nothrow` and `pure` is faked.
 private extern (C) void[] _d_arraysetlengthT(const TypeInfo ti, size_t newlength, void[]* p) nothrow pure;
 private extern (C) void[] _d_arraysetlengthiT(const TypeInfo ti, size_t newlength, void[]* p) nothrow pure;
 
+extern(C) {
+    bool gc_shrinkArrayUsed(void[] slice, size_t existingUsed, bool atomic) nothrow;
+    void[] gc_getArrayUsed(void *ptr, bool atomic) nothrow;
+
+}
+
+/**
+Shrink the "allocated" length of an array to be the exact size of the array.
+
+It doesn't matter what the current allocated length of the array is, the
+user is telling the runtime that he knows what he is doing.
+
+Params:
+    ti = `TypeInfo` of array type
+    arr = array to shrink. Its `.length` is element length, not byte length, despite `void` type
+*/
+void _d_arrayshrinkfit(T)(void[] arr) nothrow
```

Comment:
```suggestion
void _d_arrayshrinkfit(T)(T[] arr) @trusted
```
The compiler infers `pure`, `@safe` and `nothrow` for templates. But hooks are almost never safe, so mark them as `@trusted`.